compiler/hash-layout/src/constant.rs

```diff
@@ -0,0 +1,224 @@
+    pub fn usize<C: HasDataLayout>(value: u64, ctx: &C) -> Self {
+        let kind = ConstKind::Scalar(Scalar::from_usize(value, ctx));
+        Self::new(COMMON_REPR_TYS.usize, kind)
+    }
+
+    /// Attempt to coerce a [Const] into a `usize`.
+    pub fn try_to_target_usize<C: HasDataLayout>(&self, ctx: &C) -> Option<usize> {
+        if self.ty == COMMON_REPR_TYS.usize {
+            self.as_scalar().to_target_usize(ctx).try_into().ok()
+        } else {
+            None
+        }
+    }
+
+    pub fn as_big_int(&self) -> BigInt {
+        match self.kind {
+            ConstKind::Scalar(scalar) => scalar.to_big_int(self.ty.is_signed()),
+            ConstKind::Pair { data, .. } => data.to_big_int(),
+            _ => panic!("cannot cast to bigint"),
+        }
+    }
+}
+
+macro_rules! const_from_ty_impl {
+    ($($ty:ident),*) => {
+        $(
+            impl From<$ty> for Const {
+                fn from(value: $ty) -> Self {
+                    Const::new(paste!(COMMON_REPR_TYS.[<$ty>]), ConstKind::Scalar(Scalar::from(value)))
+                }
+            }
+        )*
+    };
+}
+
+const_from_ty_impl!(f32, f64, char, u8, u16, u32, u64, u128);
+
+macro_rules! try_from {
+    ($($ty:ty),*) => {
+        $(
+            impl TryFrom<&Const> for $ty {
+                type Error = Size;
+                #[inline]
+                fn try_from(value: &Const) -> Result<Self, Size> {
+                    // The `unwrap` cannot fail because to_bits (if it succeeds)
+                    // is guaranteed to return a value that fits into the size.
+                    let ConstKind::Scalar(scalar) = value.kind else {
+                        return Err(Size::ZERO) // @@Todo: change this to be more specific to the error?
+                    };
+
+                    scalar.to_bits(Size::from_bytes(std::mem::size_of::<$ty>()))
+                       .map(|u| u.try_into().unwrap())
+                }
+            }
+        )*
+    }
+}
+
+try_from!(u8, u16, u32, u64, u128);
+
+/// The kind of constants that can be represented within the Hash IR.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum ConstKind {
+    /// Exotic constant, that has no size, effectively a unit.
+    Zero,
+
+    /// A constant scalar value.
+    Scalar(Scalar),
+
+    /// A pair is only used to represent `str` constants, which is the interned
+    /// string allocation, and the length. It is equivalent to the `Abi::Pair`
+    /// representation.
+    Pair {
+        /// The data pointer of the slice. This is represented as just
+        /// being an
+        data: AllocId,
+
+        /// The length of the slice.
+        len: Scalar,
+    },
+
+    /// An allocated constant value, this is used to represent everything
+    /// that is non-scalar, including slices, ADTs and arrays.
+    Alloc {
+        /// The offset into the given allocation that is being used.
+        offset: Size,
+
+        /// The allocation itself.
+        alloc: AllocId,
+    },
```

Comment:
What is the difference between these two?

Reply:
So the only difference between `Pair` and `Alloc` is essentially that a `Pair` is used for an allocation that requires some metadata (i.e. the length of the allocation). Realistically they could be combined into one, however there is some additional logic in codegen (which involves ABI correctness and optimisations) that I don't know how to separate well without having this variant to separate.

I think leaving a `@@Todo` to figure out how to combine these two neatly is a good shout. 

For now, it should be used for strings and slices. Implementation detail from non-codegen parts will be internal - it shouldn't be concerned whether it is a pair or not.

What do you think of this?

Reply:
That sounds good; ideally strings should be able to be implemented in the language so they are all just u8 slices

Reply:
Yeah that is indeed an ideal, however the compiler at some stages will still need to treat `SizedPointer` like data specially because it can sometimes affect the ABI/codegen (in order to optimise it out).

Reply:
The alternative is to directly bake an `str` type and then represent it as an `Alloc`, i.e. `<size> <ptr_to_another_alloc>` and then the `ptr_to_another_alloc` is the underlying bytes buffer for the string.